File: aspnet-in-practice/08-09/VB/CoolMVCBlog/Utils/DependencyOutputCacheAttribute.vb
Public Class DependencyOutputCacheAttribute
    Inherits OutputCacheAttribute

    Public Property ParameterName As String
    Public Property BasePrefix As String

    Public Overrides Sub OnResultExecuting(ByVal filterContext As ResultExecutingContext)
        MyBase.OnResultExecuting(filterContext)

        Dim key As String
        If String.IsNullOrEmpty(BasePrefix) Then
            key = filterContext.RouteData.Values("action").ToString + "_" +
                                filterContext.RouteData.Values("controller").ToString
        Else
            key = BasePrefix
        End If

        If Not String.IsNullOrEmpty(ParameterName) Then
            key += "_" + filterContext.RouteData.Values(ParameterName).ToString
        End If

        filterContext.HttpContext.Cache.Insert(
            key, key, Nothing, Cache.NoAbsoluteExpiration, Cache.NoSlidingExpiration)
        filterContext.HttpContext.Response.AddCacheItemDependency(key)
    End Sub

End Class

Public Class RemoveCachedAttribute
    Inherits ActionFilterAttribute

    Public Property ParameterName As String
    Public Property BasePrefix As String

    Public Overrides Sub OnResultExecuting(ByVal filterContext As ResultExecutingContext)
        MyBase.OnResultExecuting(filterContext)

        Dim key As String
        If String.IsNullOrEmpty(BasePrefix) Then
            key = filterContext.RouteData.Values("action").ToString + "_" +
                                filterContext.RouteData.Values("controller").ToString
        Else
            key = BasePrefix
        End If

        If Not String.IsNullOrEmpty(ParameterName) Then
            key += "_" + filterContext.RouteData.Values(ParameterName).ToString
        End If

        filterContext.HttpContext.Cache.Remove(key)
    End Sub
End Class

File: aspnet-in-practice/08-09/CS/CoolMVBlog/Utils/DependencyOutputCacheAttribute.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using System.Web.Caching;

namespace CoolMvcBlog.Utils
{
    public class DependencyOutputCacheAttribute : OutputCacheAttribute
    {
        public string ParameterName { get; set; }
        public string BasePrefix { get; set; }

        public override void OnResultExecuting(ResultExecutingContext filterContext)
        {
            base.OnResultExecuting(filterContext);
            
            string key = string.IsNullOrEmpty(BasePrefix) ? 
                filterContext.RouteData.Values["action"].ToString() + "_" +
                filterContext.RouteData.Values["controller"].ToString() : BasePrefix;
            if (!string.IsNullOrEmpty(ParameterName))
                key += filterContext.RouteData.Values[ParameterName];
            
            filterContext.HttpContext.Cache.Insert(key, key, null, Cache.NoAbsoluteExpiration, Cache.NoSlidingExpiration);
            filterContext.HttpContext.Response.AddCacheItemDependency(key);
        }
    }

    public class RemoveCachedAttribute : ActionFilterAttribute
    {
        public string ParameterName { get; set; }
        public string BasePrefix { get; set; }

        public override void OnResultExecuting(ResultExecutingContext filterContext)
        {
            base.OnResultExecuting(filterContext);

            string key = string.IsNullOrEmpty(BasePrefix) ?
                filterContext.RouteData.Values["action"].ToString() + "_" +
                filterContext.RouteData.Values["controller"].ToString() : BasePrefix;

            if (!string.IsNullOrEmpty(ParameterName))
                key += filterContext.RouteData.Values[ParameterName];

            filterContext.HttpContext.Cache.Remove(key);
        }
    }
}